python/chemiscope/explore/__init__.py

```diff
@@ -124,39 +203,108 @@ def soap_kpca_featurize(frames, environments):
-        raise TypeError("'featurize' must be a callable (function)")
     if properties is None:
         properties = {}
 
-    # Apply dimensionality reduction from the provided featurizer
-    if featurize is not None:
-        X_reduced = featurize(frames, environments)
-    else:
-        # Run default featurizer
-        featurizer = PETMADFeaturizer(
-            version="latest", device=device, batch_size=batch_size
+    extracted_properties = extract_properties(frames, environments)
+    merged_properties = {**extracted_properties, **properties}
+
+    if isinstance(featurizer, str):
+        featurizer_instance = get_featurizer(featurizer, device, batch_size)
+        merged_properties["features"] = featurizer_instance(frames, environments)
+    elif callable(featurizer):
+        merged_properties["features"] = featurizer(frames, environments)
+    elif featurizer is None:
+        versions = list(FEATURIZER_VERSION_MAP.keys())
+        raise ValueError(
+            "Featurizer must be a version (string) of the default featurizer, "
+            "callable (function) or a None. Available default featurizer options are: "
+            f"{', '.join(versions) if len(versions) > 1 else versions[0]}"
         )
-        X_reduced = featurizer(frames, environments)
 
-    # Add dimensionality reduction results to properties
-    properties["features"] = X_reduced
+    if mode != "structure":
+        if _count_effective_properties(merged_properties) < 2:
+            raise ValueError(
+                "To display the widget, you need to provide at least two properties to "
+                "plot: either pass explicit properties using the 'properties' argument,"
+                " where each dimension counts as a separate property, or use a "
+                "featurizer (e.g. 'pet-mad-1.0') that generates properties (features)"
+            )
+
+        # automatically set features to axes if no map settings
+        settings = _configure_map_settings(settings, merged_properties)
 
-    return show(
+    widget = show(
         frames=frames,
-        properties=properties,
+        properties=merged_properties,
         shapes=None,
         environments=environments,
         settings=settings,
         mode=mode,
     )
+
+    if write_input is not None:
+        widget.save(write_input)
+
+    return widget
+
+
+def _count_effective_properties(properties):
+    """
+    Count the total number of properties that can we displayed in the widget,
+    where each dimension counts as a separate property
+    """
+    count = 0
+    for value in properties.values():
+        if hasattr(value, "shape") and len(value.shape) > 1:
+            # for multi-dimentional props, each will be displayed as a separate property
+            # in the widget
+            count += value.shape[1]
+        else:
+            count += 1
+    return count
+
+
+def _configure_map_settings(settings, merged_properties):
```

Comment:
why is this needed? It seems to be duplicating logic from the TypeScript code

Reply:
Ah, I would just create the properties dict by adding the features there first if we are computing them. They should end up first in the JSON input and be picked by the visualizer.